File: src/main/java/com/github/joelws/release/tracker/service/release/operation/ListReleaseServiceOperation.java
package com.github.joelws.release.tracker.service.release.operation;

import com.github.joelws.release.tracker.conversion.ReleaseToReleaseDtoAdapter;
import com.github.joelws.release.tracker.dto.release.ReleaseDto;
import com.github.joelws.release.tracker.entity.release.Release;
import com.github.joelws.release.tracker.handler.JsonResponse;
import com.github.joelws.release.tracker.service.ServiceHelper;
import com.github.joelws.release.tracker.service.ServiceOperation;
import com.github.joelws.release.tracker.service.release.execution.ListReleaseServiceExecution;

import java.util.List;
import java.util.stream.Collectors;

import javax.ws.rs.core.Response;


public class ListReleaseServiceOperation extends ServiceOperation<Void> {

    private final ServiceHelper helper;

    private final ListReleaseServiceExecution listReleaseServiceExecution;

    public ListReleaseServiceOperation(final ServiceHelper helper,
                                       final ListReleaseServiceExecution listReleaseServiceExecution) {
        this.helper = helper;
        this.listReleaseServiceExecution = listReleaseServiceExecution;
    }

    @Override
    protected Response delegate(final Void param) {
        final List<Release> result = listReleaseServiceExecution.execute(null);

        Response response;

        if (!result.isEmpty()) {
            final ReleaseToReleaseDtoAdapter adapter = helper.getFactory().getImpl(ReleaseToReleaseDtoAdapter.class);
            final List<ReleaseDto> adaptedResultList = result.stream().map(adapter::adapt).collect(Collectors.toList());
            response = helper.getRestResponseBuilder().build(200, adaptedResultList);
        } else {
            response = helper.getRestResponseBuilder().build(new JsonResponse(404, "No releases exist"));
        }
        return response;
    }
}

File: src/main/kotlin/com/github/joelws/release/tracker/service/release/operation/ListReleaseServiceOperation.kt
package com.github.joelws.release.tracker.service.release.operation

import com.github.joelws.release.tracker.conversion.ReleaseAdapter
import com.github.joelws.release.tracker.entity.release.Release
import com.github.joelws.release.tracker.interfaces.Adapter
import com.github.joelws.release.tracker.model.release.ReleaseModel
import com.github.joelws.release.tracker.response.RestResponse.NotFound
import com.github.joelws.release.tracker.response.RestResponse.SuccessWithEntity
import com.github.joelws.release.tracker.response.build
import com.github.joelws.release.tracker.service.ServiceExecution
import com.github.joelws.release.tracker.service.ServiceHelper
import com.github.joelws.release.tracker.service.ServiceOperation
import javax.ws.rs.core.Response


open class ListReleaseServiceOperation(private val helper: ServiceHelper,
                                       private val listReleaseServiceExecution: ServiceExecution<Nothing?, List<Release>>) : ServiceOperation<Nothing?> {

    override fun delegate(param: Nothing?): Response {

        val resultList = listReleaseServiceExecution.execute(null)

        return if (resultList.isNotEmpty()) {
            @Suppress("UNCHECKED_CAST")
            val releaseAdapter: Adapter<Release, ReleaseModel> = helper.adapterFactory.getAdapter(ReleaseAdapter::class.java) as Adapter<Release, ReleaseModel>
            val adaptedResultList = resultList.map { releaseAdapter.adapt(it) }
            SuccessWithEntity(adaptedResultList).build()
        } else {
            NotFound("No releases exist").build()
        }
    }
}
